src/components/BarChart/BarChart.tsx

```diff
@@ -0,0 +1,238 @@
+        height: 300,
+        minWidth: 300,
+        maxWidth: 1000,
+        isResponsive: false,
+        onClick: () => { return; }
+    };
+
+    public refs: {
+        [argument: string]: (Element);
+        container: (HTMLInputElement);
+    };
+
+    private margin = { top: 50, right: 50, bottom: 50, left: 50 };
+    private width = this.props.width - this.margin.right - this.margin.left;
+    private height = this.props.height - this.margin.top - this.margin.bottom;
+    private x;
+    private y;
+    private focus;
+    private svg;
+    private container;
+
+    constructor(props: IBarChartProps) {
+        super(props);
+
+        this.x = this.generateX();
+        this.y = this.generateY();
+
+        // if (this.props.isResponsive) {
+        //     this.width = window.innerWidth / 2;
+        //     this.height = 0.7 * this.width;
+        // }
+    }
+
+    public componentWillReceiveProps(nextProps: IBarChartProps) {
+        // console.log('next props', nextProps);
+
+        // see what props have changed and set appropriate enum based on that
+    }
+
+    public shouldComponentUpdate(nextProps: IBarChartProps, nextState: null) {
+        if (this.props.id !== nextProps.id ||
+            this.props.width !== nextProps.width ||
+            this.props.height !== nextProps.height ||
+            this.props.barColor !== nextProps.barColor) {
+                return true;
+            }
+        if (!this.arraysEqual(this.props.data, nextProps.data)) { return true; }
+        return false;
+    }
+
+    private arraysEqual(arr1: IBarChartData[], arr2: IBarChartData[]) {
+        if (arr1.length !== arr2.length) { return false; }
+        for ( let i = arr1.length; i--; ) {
+            if (!this.compareValues(arr1[i], arr2[i])) { return false; }
+        }
+        return true;
+    }
+
+    private compareValues(data1: IBarChartData, data2: IBarChartData) {
+        if (typeof data1.argument !== 'number' && typeof data1.argument !== 'string') {
+            if ((data1.argument as Date).getTime() !== (data2.argument as Date).getTime()) { return false; }
+        }
+        if (data1.argument !== data2.argument) { return false; }
+        if (data1.frequency !== data2.frequency) { return false; }
+        return true;
+    }
+
+    public componentDidMount() {
+        this.init();
+        if (this.props.isResponsive) { window.addEventListener('resize', () => this._onResize()); }
+    }
+
+    public componentDidUpdate() {
+        this.redraw();
+        // console.log('updating bar chart...'); // debug...
+    }
+
+    public render() {
+        return <div className={'bar-chart-component'} ref={'container'}></div>;
+    }
+
+    private init() {
+        const xAxisClass = classNames('x-axis', this.props.id);
+        const yAxisClass = classNames('y-axis', this.props.id);
+
+        this.container = this.createContainer();
+        this.container.append('g').attr('class', xAxisClass).call(this.generateXAxis())

```

Comment:
can you add some comments as to what is happening in this code. The d3 is powerful, but can be cryptic sometimes xD